bftengine/src/communication/TlsTCPCommunication.cpp

```diff
@@ -245,418 +219,493 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+      LOG_ERROR(_logger,
+                "close_socket, exception, node " << _selfId << ", "
+                                                               "dest: "
+                                                 << _destId << ", connected: "
+                                                 << _connected
+                                                 << ", ex: " << e.what());
     }
 
-    LOG_TRACE("exit, node " << _selfId << ", dest: " << _destId
-                            << ", connected: " << connected << ", closed: "
-                            << _closed);
+    get_socket().close(ec);
+    was_error(ec, "close_socket");
   }
 
+  /**
+   * this method closes the socket and frees the object by calling the _fOnError
+   */
   void close() {
-    _connecting = true;
-    LOG_TRACE("enter, node " << _selfId << ", dest: " << _destId
-                             << ", connected: " << connected << ", closed: "
-                             << _closed);
+    if (_closed)
+      return;
 
-    lock_guard<recursive_mutex> lock(_connectionsGuard);
+    set_authenticated(false);
+    set_connected(false);
+    set_closed(true);
+
+    LOG_DEBUG(_logger, "close, node " << _selfId << ", dest: " << _destId
+                                      << ", connected: " << _connected
+                                      << ", closed: " << _closed);
 
-    connected = false;
-    _closed = true;
     _connectTimer.cancel();
 
+    B_ERROR_CODE ec;
     try {
-      B_ERROR_CODE ec;
-      get_socket().shutdown(boost::asio::ip::tcp::socket::shutdown_both, ec);
-      get_socket().close();
-      if (_pSslContext)
-        _pSslContext.release();
+      if (_socket) {
+        _socket->shutdown();
+        get_socket().shutdown(
+            boost::asio::ip::tcp::socket::shutdown_both, ec);
+        was_error(ec, "close,shutdown");
+      }
     } catch (std::exception &e) {
-      LOG_ERROR("exception, node " << _selfId << ", dest: " << _destId
-                                   << ", connected: " << connected
-                                   << ", ex: " << e.what());
+      LOG_ERROR(_logger, "exception, node " << _selfId << ", dest: " << _destId
+                                            << ", connected: " << _connected
+                                            << ", ex: " << e.what());
     }
 
-    LOG_TRACE("exit, node " << _selfId << ", dest: " << _destId
-                            << ", connected: " << connected << ", closed: "
-                            << _closed);
-
-    _fOnError(_destId);
-  }
-
-  bool was_error(const B_ERROR_CODE &ec, string where) {
-    if (ec) LOG_ERROR("where: " << where << ", node " << _selfId << ", dest: "
-                                << _destId << ", connected: " << connected
-                                << ", ex: " << ec.message());
+    get_socket().close(ec);
+    was_error(ec, "close");
 
-    return (ec != 0);
+    if (_fOnError) {
```

Comment:
Yeah it's kind of confusing. I usually check the pointers before using but in this case it makes sense to remove this check so it will become clear that the pointer should always be valid.